src/core/segment_reader.rs

```diff
@@ -330,6 +432,64 @@ impl SegmentReader {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+/// FieldMetadata
+pub struct FieldMetadata {
+    /// The field name
+    // Notice: Don't reorder the declaration of 1.field_name 2.typ, as it is used for ordering by
+    // field_name then typ.
+    pub field_name: String,
+    /// The field type
+    // Notice: Don't reorder the declaration of 1.field_name 2.typ, as it is used for ordering by
+    // field_name then typ.
+    pub typ: Type,
+    /// Is the field indexed for search
+    pub indexed: bool,
+    /// Is the field stored in the doc store
+    pub stored: bool,
+    /// Is the field stored in the columnar storage
+    pub fast: bool,
+}
+impl BitOrAssign for FieldMetadata {
+    fn bitor_assign(&mut self, rhs: Self) {
+        assert!(self.field_name == rhs.field_name);
+        assert!(self.typ == rhs.typ);
+        self.indexed |= rhs.indexed;
+        self.stored |= rhs.stored;
+        self.fast |= rhs.fast;
+    }
+}
+
+/// Helper to merge the field metadata from multiple segments.
+pub fn merge_field_meta_data(
+    input: Vec<Vec<FieldMetadata>>,
+    schema: &Schema,
+) -> Vec<FieldMetadata> {
+    // Maybe too slow for the high cardinality case
+    fn is_field_stored(field_name: &str, schema: &Schema) -> bool {
```

Comment:
nitpick: Can we have the function outside of this?

I personally find it less readable.
I assume your point is that you want to namespace the function as much as possible.
One trouble with these is that it makes it non trivial to know whether they capture their environment or not.

So answering the question, "what does thjis function does" actually needs to look at its impl, as opposed to looking at the prototype of the function. It could hide any surprises.

Reply:
I think `fn` can't capture anything?
I moved it outside